pallets/gear/src/tests.rs

```diff
@@ -3703,6 +3705,92 @@ fn cascading_messages_with_value_do_not_overcharge() {
     });
 }
 
+#[test]
+fn execution_over_blocks() {
+    init_logger();
+    new_test_ext().execute_with(|| {
+        use demo_calc_hash::{Method, Package};
+        use demo_calc_hash_aggregator::WASM_BINARY;
+        use demo_calc_hash_calculator::WASM_BINARY as CALC_WASM_BINARY;
+        let block_gas_limit = BlockGasLimitOf::<Test>::get();
+        let max_gas_spent_per_hash = 30_000_000_000u64;
+
+        // deply calculator
+        assert_ok!(Gear::submit_program(
+            Origin::signed(USER_1),
+            CALC_WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            EMPTY_PAYLOAD.to_vec(),
+            block_gas_limit,
+            0,
+        ));
+        let calculator = get_last_program_id();
+
+        // deploy aggregator
+        assert_ok!(Gear::submit_program(
+            Origin::signed(USER_1),
+            WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            (max_gas_spent_per_hash, calculator).encode(),
+            block_gas_limit,
+            0,
+        ));
+        let aggregator = get_last_program_id();
+
+        program_exists(calculator.into_origin());
+        program_exists(aggregator.into_origin());
+
+        // the result of running sha2_256 on [0; 32] itself for 42 times
+        //
+        // 0 -> [0; 32]
+        // 1 -> sha2_256(&[0; 32])
+        // 2 -> sha2_256(sha2_256(&[0; 32]))
+        // 3 -> ...
+        let expected = [
+            138, 64, 61, 254, 55, 192, 102, 102, 40, 78, 11, 110, 11, 207, 223, 93, 190, 11, 114,
+            133, 69, 24, 177, 45, 247, 184, 60, 142, 126, 223, 94, 220,
+        ];
+
+        // trigger calculation
+        assert_ok!(Gear::send_message(
+            Origin::signed(USER_1),
+            aggregator,
+            Method::Start(Package {
+                paths: vec![[0; 32]],
+                expected,
+            })
+            .encode(),
+            block_gas_limit,
+            0,
+        ));
+
+        run_to_block(2, None);
+        assert_eq!(maybe_last_mail(USER_1), None);
+
+        // loop calculation
+        let path: Vec<[u8; 32]>;
+        loop {
```

Comment:
I prefer `while maybe_last_mail.is_none`

Some counters and debug outputs to show that this runs multiple times would also be informative